### ISSUE TITLE
Fine-tuning

### ISSUE THREAD

**Issue by jessiclassy** [AnanthaR20/ling573#21]
Using the HuggingFace tutorial [here](https://huggingface.co/docs/transformers/en/tasks/summarization), attempt to fine-tune PEGASUS on BillSum exactly as-is. We want to make sure we have working scripts that we can reuse with simplified summary data. 

Model options:
- https://huggingface.co/google/pegasus-large
- https://huggingface.co/zphang/pegasus-x-base

System options:
- Patas/Condor
- Hyak

**BEST PRACTICES**
- Make sure that runtimes are logged/recorded somewhere (ex: Condor logs) or else generate those timestamps yourself 
- Make sure to account for the edge case that the process exits abnormally (ex: do we want to write the partially fine-tuned model parameters to a checkpoint or something?)

**Comment by jessiclassy** [comment]
Dumping out my notes/adding links from my tabs to iron out the core Patas finetune setup (before adjusting hyperparams):
- Per source paper (Zhang et al. 2020), we need at least `label_smoothing_factor=0.1` but to help with regularization we might want to bump it up to 0.15 or even 0.2
- Source paper learning rate is 2e-4 but their batch size is 256, so we probably need to adjust the learning rate to fit our comparatively tiny batch size...maybe to 2e-5 or even 2e-6
- We can also try tweaking the batch size up as Patas memory allows. We could use [LoRa](https://huggingface.co/blog/mlabonne/sft-llama3) and potentially get the batch size back into the typical 16-32 range (at least from what I've skimmed online). [Here](https://www.philschmid.de/fine-tune-flan-t5-peft) is another Seq2Seq LoRa tutorial, I think we should follow this pretty closely! Note that LoRa trades ~30ish% runtime increase for ~30% memory efficiency
- [This](https://arxiv.org/abs/2307.13831) paper suggests that we need higher `max_steps` than 100000 (from the source paper) given our batch size limitations...but [this article](https://medium.com/@liana.napalkova/fine-tuning-small-language-models-practical-recommendations-68f32b0535ca) suggests that finetuning is completed in 1-3 epochs?
- Docs don't make it clear that the param `do_predict=True` will do anything...
- We could just add a final line: `model.predict(tokenized_billsum["test"])` to the code. source [here](https://docs.pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html#the-train-function)
- Interestingly, we don't have `do_train=True` at the moment so it looks like this is instead handled by `trainer.train()` I guess? source [here](https://docs.pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html#the-train-function)


From PEGASUS paper (Zhang et al. 2020):
> Hyperparamters of the pre-training and fine-tuning stages reported in section 6...of fine-tuning PEGASUSLARGE were decided by grid search while others were decided by empirically default commonly used values

Potential resource for hyperparameter tuning:
- https://docs.pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html

**Comment by jessiclassy** [comment]
Independently of these hyperparameter/finetuning details, we need to adjust `finetune_model.py` to accept one/more preprocessing components on the dataset before tokenizing. 
- We could load configuration settings from a file/set of files
- We could use `argparse` to accept command line flags

(we probably need to sketch this out on a whiteboard)
